src/transformers/models/realm/retrieval_realm.py

```diff
@@ -0,0 +1,129 @@
+# coding=utf-8
+# Copyright 2020, The RAG Authors and The HuggingFace Inc. team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Realm Retriever model implementation."""
+import numpy as np
+
+# from ...tokenization_utils_base import BatchEncoding
+from ...utils import logging
+
+
+logger = logging.get_logger(__name__)
+
+
+def convert_tfrecord_to_np(block_records_path, num_block_records):
+    import tensorflow.compat.v1 as tf
+
+    blocks_dataset = tf.data.TFRecordDataset(block_records_path, buffer_size=512 * 1024 * 1024)
+    blocks_dataset = blocks_dataset.batch(num_block_records, drop_remainder=True)
+    np_record = next(blocks_dataset.take(1).as_numpy_iterator())
+
+    return np_record
+
+
+class ScaNNSearcher:
+    def __init__(
+        self,
+        db,
+        num_neighbors,
+        dimensions_per_block=2,
+        num_leaves=1000,
+        num_leaves_to_search=100,
+        training_sample_size=100000,
+    ):
+        """Build scann searcher."""
+
+        from scann.scann_ops.py.scann_ops_pybind import builder as Builder
+
+        builder = Builder(db=db, num_neighbors=num_neighbors, distance_measure="dot_product")
+        builder = builder.tree(
+            num_leaves=num_leaves, num_leaves_to_search=num_leaves_to_search, training_sample_size=training_sample_size
+        )
+        builder = builder.score_ah(dimensions_per_block=dimensions_per_block)
+
+        self.searcher = builder.build()
+
+    def search_batched(self, question_projection):
+        retrieved_block_ids, _ = self.searcher.search_batched(question_projection.detach().cpu())
+        # Must return cpu tensor for subsequent numpy operations
+        #        return torch.tensor(retrieved_block_ids.astype("int64"), device=torch.device("cpu"))
+        return retrieved_block_ids.astype("int64")
+
+
+class RealmRetriever:
```

Comment:
We will make this prettier iteratively from now. @lhoestq - currently we are passing the `block_records_path` which is a long to a `.tfr` file, *e.g.*: https://storage.cloud.google.com/orqa-data/enwiki-20181220/blocks.tfr . Could we add such a file format to `datasets` somehow?

Reply:
@patrickvonplaten Should we handle `block_emb` tensor along with `block_records` in  `datasets` as well or at least provide a warning in the model card or docs that a checkpoint is corresponded to a specific `block_records` as they are bound together.

Reply:
Good point! I think we have to store the retrival `block_records` in the same model repo as we store the PyTorch model that contains the `block_emb` so that it is crystal clear that one should use those two together. The final API should look as follows:

```python
model_id = "qqaatw/realm-open-qa"

tokenizer = RealmTokenizer.from_pretrained(model_id)
retriever = RealmRetriever.from_pretrained(model_id)

model = RealmForOpenQA.from_pretrained(model_id, retriever)

question = "Who is the pioneer in modern computer science?"

question = tokenizer(question, padding=True, truncation=True, max_length=model.config.searcher_seq_len)

predicted_answer_ids = model(**question).predicted_answer
predicted_answer = tokenizer.decode(predicted_answer_ids)
```

Reply:
The API looks great! So it seems that we don't need `datasets` as the `block_records` is stored in the model repo, and we might need to override `retriever.from_pretrained` in order to load such a `.tfr` file? I think `PreTrainedModel.from_pretrained` cannot handle this file right?

Reply:
That's a good point here actually! Let me sync quickly with @lhoestq on this. We might be able to go for a much simpler solution as you suggested. It would actually be ideal if we could full remove the tf dependency here. Let's see, I'll keep you updated